src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ILScanner.cs

```diff
@@ -467,13 +468,31 @@ public ScannedDevirtualizationManager(ImmutableArray<DependencyNodeCore<NodeFact
                                 // Record all interfaces this class implements to _interfaceImplementators
                                 foreach (DefType baseInterface in type.RuntimeInterfaces)
                                 {
-                                    if (CanAssumeWholeProgramViewOnInterfaceUse(baseInterface))
+                                    if (CanAssumeWholeProgramViewOnInterfaceUse(factory, type, baseInterface))
                                     {
                                         RecordImplementation(baseInterface, type);
                                     }
                                 }
                             }
 
+                            // Interfaces implemented by arrays and array enumerators have weird casting rules
+                            // due to array covariance (string[] castable to object[], or int[] castable to uint[]).
+                            // Disqualify such interfaces.
+                            if (!type.IsCanonicalSubtype(CanonicalFormKind.Any) &&
```

Comment:
Is the `IsCanonicalSubtype` early out correct? Consider:
```
var a = new int[1];
var b = (X<string>.MyEnum[])(object)a;

Console.WriteLine(b);

class X<T>
{
    public enum MyEnum
    {
    }
}
```

Reply:
It was fine for that - we were disqualifying template types above (IsCanonical detects template types).

I moved these two checks together. It looks better.